### PR TITLE
Add support for empty Lists and empty Maps

### PR DESCRIPTION
**Pull request by pablocasares**
# TL;DR
Add support for empty Lists and empty Maps in SdkBindingData. 
Add helper methods to create collections of most common types:
- ofStringCollection
- ofFloatCollection
- ofIntegerCollection
- ofBooleanCollection
- ofDurationCollection
- ofDatetimeCollection

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Added new helpers in SdkBindingData class and modified methods ofMap and ofCollection to ask the user to pass the LiteralType in order to be able to know the underlying type of the collection/map.

### CODE REVIEW

**Review by andresgomezfrr — APPROVED** [review]

File: flytekit-examples/src/main/java/org/flyte/examples/PhoneBookWorkflow.java [hunk 1]
```diff
@@ -51,7 +51,7 @@ public abstract static class Output {
      */
     public static PhoneBookWorkflow.Output create(List<String> phoneNumbers) {
       return new AutoValue_PhoneBookWorkflow_Output(
-          SdkBindingData.ofCollection(phoneNumbers, SdkBindingData::ofString));
+          SdkBindingData.ofStringCollection(phoneNumbers));
```
Comment (narape) [inline flytekit-examples/src/main/java/org/flyte/examples/PhoneBookWorkflow.java:54]: Nice that you retrofitted the new methods on old calls

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkBindingData.java [hunk 2]
```diff
@@ -88,63 +88,134 @@ private static <T> SdkBindingData<T> ofPrimitive(Function<T, Primitive> toPrimit
...
   public static <T> SdkBindingData<List<T>> ofCollection(
-      List<T> collection, Function<T, SdkBindingData<T>> function) {
+      List<T> collection, LiteralType literalType, Function<T, SdkBindingData<T>> function) {
     return SdkBindingData.ofBindingCollection(
-        collection.stream().map(function).collect(Collectors.toList()));
+        literalType, collection.stream().map(function).collect(Collectors.toList()));
+  }
+
+  private static <T> SdkBindingData<List<T>> ofTCollection(
+      List<T> collection, LiteralType literalType, Function<T, BindingData> bindingDataFn) {
```
Comment (narape) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkBindingData.java:97]:
```suggestion
  interface ToBindingDataFunction<T> {
    BindingData apply(T value);
  }
      List<T> collection, LiteralType literalType, ToBindingDataFunction bindingDataFn) {
```
```suggestion
  private static <T> SdkBindingData<List<T>> ofTCollection(
```
You had to change the name of the method because its eraused firm is the same as the other. But you gave it a name that it is applicable to both methods, both receive a `List<T>`

I think that if we introduce our own interface we can break the erauser problem and keep using the `ofCollection` name

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkBindingData.java [hunk 3]
```diff
@@ -88,63 +88,134 @@ private static <T> SdkBindingData<T> ofPrimitive(Function<T, Primitive> toPrimit
...
     return SdkBindingData.ofBindingMap(
+        literalType,
         map.entrySet().stream()
             .map(e -> Map.entry(e.getKey(), bindingFunction.apply(e.getValue())))
             .collect(toMap(Map.Entry::getKey, Map.Entry::getValue)));
   }
 
-  public static SdkBindingData<Map<String, Double>> ofFloatMap(Map<String, Double> map) {
-    return ofMap(map, SdkBindingData::ofFloat);
+  private static <T> SdkBindingData<Map<String, T>> ofTMap(
```
Comment (narape) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkBindingData.java:156]: Same comment here, we can reuse the same interface

File: flytekit-java/src/test/java/org/flyte/flytekit/SdkBindingDataTest.java [hunk 4]
```diff
@@ -60,15 +65,126 @@ public void testOfBindingCollection() {
 
   @Test
   public void testOfBindingCollection_empty() {
-    IllegalArgumentException e =
-        assertThrows(
-            IllegalArgumentException.class, () -> SdkBindingData.ofBindingCollection(emptyList()));
+    List<BindingData> expected = emptyList();
```
Comment (narape) [inline flytekit-java/src/test/java/org/flyte/flytekit/SdkBindingDataTest.java:68]: Nice

File: flytekit-java/src/test/java/org/flyte/flytekit/SdkBindingDataTest.java [hunk 5]
```diff
@@ -60,15 +65,126 @@ public void testOfBindingCollection() {
...
+    SdkBindingData<List<Long>> output =
+        SdkBindingData.create(
+            BindingData.ofCollection(expected),
+            LiteralType.ofCollectionType(LiteralTypes.INTEGER),
+            emptyList());
 
     assertThat(
-        e.getMessage(),
+        output,
```
Comment (narape) [inline flytekit-java/src/test/java/org/flyte/flytekit/SdkBindingDataTest.java:76]: These assertions are upside down, we assume that the create method from autovalue works correctly so what we really want to test here is the 
```java
            SdkBindingData.ofBindingCollection(
                LiteralType.ofCollectionType(LiteralType.ofSimpleType(SimpleType.INTEGER)),
                emptyList())));
```
So that is what should be as the first parameter of the `assertThat`
So 
* rename `output` to `expected`
* what you have inside of the `equalsTo` name it as `output` before the whole `assertThat`
* Then the assertThat should be 
```java
    assertThat(output, equalTo(expected));
```

Comment (narape) [reply flytekit-java/src/test/java/org/flyte/flytekit/SdkBindingDataTest.java:76]: The same applies to other tests

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkBindingData.java [hunk 6]
```diff
@@ -153,8 +153,8 @@ public static <T> SdkBindingData<Map<String, T>> ofMap(
             .collect(toMap(Map.Entry::getKey, Map.Entry::getValue)));
   }
 
-  private static <T> SdkBindingData<Map<String, T>> ofTMap(
-      Map<String, T> map, LiteralType literalType, Function<T, BindingData> bindingDataFn) {
+  private static <T> SdkBindingData<Map<String, T>> ofMapFromBindingData(
```
Comment (narape) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkBindingData.java:156]: Name it `createMap` here

File: flytekit-java/src/test/java/org/flyte/flytekit/SdkBindingDataTest.java [hunk 7]
```diff
@@ -86,15 +84,16 @@ public void testOfStringCollection() {
     input.add(SdkBindingData.ofString("1"));
     input.add(SdkBindingData.ofString("2"));
 
-    List<String> expected = new ArrayList<>();
-    expected.add("1");
-    expected.add("2");
+    List<String> expectedValue = new ArrayList<>();
+    expectedValue.add("1");
+    expectedValue.add("2");
```
Comment (narape) [inline flytekit-java/src/test/java/org/flyte/flytekit/SdkBindingDataTest.java:89]:
```suggestion
    List<String> expectedValue = List.of("1", "2");
```
I think that we are in java 11, so we can use this now

Comment (narape) [reply flytekit-java/src/test/java/org/flyte/flytekit/SdkBindingDataTest.java:89]: here and the rests of the tests

File: flytekit-java/src/test/java/org/flyte/flytekit/SdkBindingDataTest.java [hunk 8]
```diff
@@ -233,14 +239,15 @@ public void testOfStringMap() {
     input.put("a", SdkBindingData.ofString("1"));
     input.put("b", SdkBindingData.ofString("2"));
 
-    Map<String, String> expected = new HashMap<>();
-    expected.put("a", "1");
-    expected.put("b", "2");
+    Map<String, String> expectedValue = new HashMap<>();
```
Comment (narape) [inline flytekit-java/src/test/java/org/flyte/flytekit/SdkBindingDataTest.java:242]: We can use `Map.of` instead of the `HashMap`s
